cookbooks/kinetic_driving_force/kinetic-driving-force.cc

```diff
@@ -0,0 +1,545 @@
+          alpha_b = profile.get_data_component(profile_pos, alpha_b_idx);
+          beta_a  = profile.get_data_component(profile_pos, beta_a_idx);
+          beta_b  = profile.get_data_component(profile_pos, beta_b_idx);
+          cp_a    = profile.get_data_component(profile_pos, cp_a_idx);
+          cp_b    = profile.get_data_component(profile_pos, cp_b_idx);
+          dG      = profile.get_data_component(profile_pos, dG_idx);
+          dS      = profile.get_data_component(profile_pos, dS_idx);
+          dV      = profile.get_data_component(profile_pos, dV_idx);
+
+          // Calculate viscosity temperature factor
+          double visc_temperature_dependence =
+            std::max(std::min(std::exp(-thermal_viscosity_exponent *
+                                       (T - T_adiabatic) / T_adiabatic),
+                              1e3),
+                     1e-3);
+          if (std::isnan(visc_temperature_dependence))
+            visc_temperature_dependence = 1.0;
+
+          // Calculate viscosity depth factor
+          double visc_depth_dependence = viscosity_prefactors[0];
+          for (unsigned int z = 0; z < transition_depths.size(); ++z)
+            {
+              const Point<dim, double> transition_point =
+                this->get_geometry_model().representative_point(
+                  transition_depths[z]);
+              double transition_pressure =
+                this->get_adiabatic_conditions().pressure(transition_point);
+              if (P_adiabatic > transition_pressure)
+                visc_depth_dependence = viscosity_prefactors[z + 1];
+            }
+
+          // Calculate viscosity
+          const double eta =
+            viscosity * visc_temperature_dependence * visc_depth_dependence;
+          const double eta_effective =
+            std::min(std::max(eta, minimum_viscosity), maximum_viscosity);
+
+          // Clamp composition to valid range
+          const double X_clamped = std::max(0.0, std::min(1.0, X));
+
+          // Average material properties
+          std::vector<double> mass_fractions(
+            this->introspection().n_compositional_fields, 1.0);
+          mass_fractions[0] = 1 - X_clamped;
+          mass_fractions[1] = X_clamped;
+
+          std::vector<double> densities(
+            this->introspection().n_compositional_fields, rho_a);
+          densities[0] = rho_a;
+          densities[1] = rho_b;
+
+          std::vector<double> alphas(
+            this->introspection().n_compositional_fields, alpha_a);
+          alphas[0] = alpha_a;
+          alphas[1] = alpha_b;
+
+          std::vector<double> betas(
+            this->introspection().n_compositional_fields, beta_a);
+          betas[0] = beta_a;
+          betas[1] = beta_b;
+
+          std::vector<double> cps(this->introspection().n_compositional_fields,
+                                  cp_a);
+          cps[0] = cp_a;
+          cps[1] = cp_b;
```

Comment:
If all you do with `rho_a,rho_b,alpha_a,...` is to fill it into these vectors, I would say move all of this code up, do not create the variables, and instead will the vectors directly:
```
std::vector<double> densities(this->introspection().n_compositional_fields);
densities[0] = profile.get_data_component(profile_pos, rho_a_idx);
densities[1] = profile.get_data_component(profile_pos, rho_b_idx);
```
and the same for the other properties.